src/components/Toolbar.tsx

```diff
@@ -0,0 +1,371 @@
+   */
+  private _renderRepoMenu() {
+    return (
+      <div className={toolbarMenuWrapperClass}>
+        <button
+          className={toolbarMenuButtonClass}
+          title={`Current repository: ${this.state.repository}`}
+          onClick={this._onRepositoryClick}
+        >
+          <span
+            className={classes(
+              toolbarMenuButtonIconClass,
+              repoIconClass,
+              'jp-Icon-16'
+            )}
+          />
+          <div className={toolbarMenuButtonTitleWrapperClass}>
+            <p className={toolbarMenuButtonTitleClass}>Current Repository</p>
+            <p className={toolbarMenuButtonSubtitleClass}>
+              {PathExt.basename(this.state.repository)}
+            </p>
+          </div>
+          <span
+            className={classes(
+              toolbarMenuButtonIconClass,
+              this.state.repoMenu ? closeMenuIconClass : openMenuIconClass,
+              'jp-Icon-16'
+            )}
+          />
+        </button>
+        {this.state.repoMenu ? null : null}
+      </div>
+    );
+  }
+
+  /**
+   * Renders a branch menu.
+   *
+   * @returns fragment
+   */
+  private _renderBranchMenu() {
+    if (!this.state.repository) {
+      return null;
+    }
+    return (
+      <div className={toolbarMenuWrapperClass}>
+        <button
+          className={toolbarMenuButtonClass}
+          title={`Change the current branch: ${this.state.branch}`}
+          onClick={this._onBranchClick}
+        >
+          <span
+            className={classes(
+              toolbarMenuButtonIconClass,
+              branchIconClass,
+              'jp-Icon-16'
+            )}
+          />
+          <div className={toolbarMenuButtonTitleWrapperClass}>
+            <p className={toolbarMenuButtonTitleClass}>Current Branch</p>
+            <p className={toolbarMenuButtonSubtitleClass}>
+              {this.state.branch}
+            </p>
+          </div>
+          <span
+            className={classes(
+              toolbarMenuButtonIconClass,
+              this.state.branchMenu ? closeMenuIconClass : openMenuIconClass,
+              'jp-Icon-16'
+            )}
+          />
+        </button>
+        {this.state.branchMenu ? (
+          <BranchMenu
+            model={this.props.model}
+            branching={this.props.branching}
+          />
+        ) : null}
+      </div>
+    );
+  }
+
+  /**
+   * Adds model listeners.
+   */
+  private _addListeners(): void {
+    // When the repository changes, we're likely to have a new set of branches:
+    this.props.model.repositoryChanged.connect(this._syncState, this);
+
+    // When the HEAD changes, decent probability that we've switched branches:
+    this.props.model.headChanged.connect(this._syncState, this);
+
+    // When the status changes, we may have checked out a new branch (e.g., via the command-line and not via the extension):
+    this.props.model.statusChanged.connect(this._syncState, this);
```

Comment:
`statusChanged` should not be listened by this component

Reply:
Removed `repositoryChanged` listener, but kept `statusChanged` listener. In short, a `headChanged` signal is not guaranteed when switching branches (e.g., via the command-line outside of the extension) and `statusChanged` catches repository changes and status changes occurring outside the extension. Yes, this might mean some "false positive" signals (e.g., checking out a previous commit triggering a `statusChanged` signal), but these should not trigger a re-render as component state will remain the same.